designs/hosted_cluster_vpc_endpoints.md

```diff
@@ -0,0 +1,121 @@
+# Create VPC Endpoints for Private Hosted Clusters
+
+Author: @mjlshen @aliceh
+
+Last Updated: 04/13/2023
+
+## Summary
+
+`aws-vpce-operator` (AVO) needs to be able to manage the lifecycle of one VpcEndpoint CR per private HyperShift cluster to enable SRE backplane access.
+
+### Current State
+
+There is no automation creating AWS VPC Endpoints to enable backplane connectivity to every private hosted cluster's AWS VPC Endpoint Service.
+
+## Motivation
+
+This change is important because:
+
+- Without it, SREs will not be able to access any private hosted cluster's API server
+
+## Relevant Stories
+
+- [SDE-1592](https://issues.redhat.com/browse/SDE-1592) Multi-tenant control plane account hierarchy
+- [OSD-15666](https://issues.redhat.com/browse/OSD-15666) Create AVO controller to manage backplane VPC Endpoints
+
+## Goals
+
+- Create a new `aws-vpce-operator` controller and CustomResourceDefinition named `vpcendpointtemplate` to manage the lifecycle of one VpcEndpoint CR per private HyperShift cluster to enable SRE backplane access.
+
+## Non-Goals/Future Work
+
+- Modify the `hostedcluster`'s public or private endpoint setting itself.
+
+## Proposal
+
+- Upon receiving an event for a private hostedcluster creation or change, template out an AVO `VpcEndpoint` CR to enable AVO's existing `vpcendpoint` controller to reconcile on.
+- Likewise, handle the deletion of this AVO `VpcEndpoint` CR when a private hostedcluster gets deleted or changes to public.
+
+- How do we figure out if a hostedcluster is private?
+  - .spec.platform.aws.endpointAccess: Private on the `hostedcluster` CR
+- How do we figure out if a hostedcluster has additionalAllowedPrincipals?
+  - .spec.platform.aws.additionalAllowedPrincipals on the `hostedcluster` CR
+- How do we figure out the VPC Endpoint Service?
+  - .status.endpointServiceName on the `awsendpointservice` CR
+
+### VpcEndpointTemplate CR
+
+```yaml
+apiVersion: avo.openshift.io/v1alpha2
+kind: VpcEndpointTemplate
+metadata:
+  name: private-hostedclusters
+  namespace: openshift-aws-vpce-operator
+spec:
+  template:
+    hostedclusterRef: ${hostedcluster_name} # It would be nice if we could get awsendpointservice name and namespace from the hostedcluster status field? Ask HOSTEDCP
```

Comment:
`hostedclusterRef` and `awsendpointserviceRef` aren't as straightforward as we would like since they are in separate namespaces. It's potentially possible, but just not super great